File: web_extras/routers/vault.py
"""Read-only HTTP surface over the secrets vault.

- ``GET /api/vault/status`` — returns availability for every well-known
  key plus any name declared by a registered domain pack via
  :meth:`~backend.core.domains.base.DomainPack.auth_vault_keys`
  (``env`` / ``keychain`` / ``missing``). Values are NEVER echoed.

The cockpit uses this to render per-pack auth badges (e.g. council
voices show "anthropic: keychain" or "openai: missing") and SMTP keys.
"""

from __future__ import annotations

from typing import Any

from fastapi import APIRouter

from backend.core.domains.registry import all_packs
from backend.core.vault import KNOWN_KEYS, status_for_keys

router = APIRouter(prefix="/api/vault", tags=["vault"])


def _merged_vault_key_order() -> list[str]:
    """``KNOWN_KEYS`` first, then any pack-declared keys not yet listed."""

    out: list[str] = []
    seen: set[str] = set()
    for k in KNOWN_KEYS:
        if k not in seen:
            out.append(k)
            seen.add(k)
    for pack in all_packs():
        for k in pack.auth_vault_keys():
            if k not in seen:
                out.append(k)
                seen.add(k)
    return out


@router.get("/status")
async def status() -> dict[str, Any]:
    refs = status_for_keys(_merged_vault_key_order())
    return {
        "ok": True,
        "count": len(refs),
        "keys": [
            {"key": r.key, "source": r.source, "available": r.available}
            for r in refs
        ],
    }

File: experiments/neural-showcase-v3/src/lib/vault.ts
/**
 * Client + hook for the TARS secrets vault.
 *
 * Read-only; values are NEVER returned by the server. The hook lets
 * the cockpit render per-pack auth badges (e.g. "anthropic: keychain"
 * vs "openai: missing").
 */

import { useCallback, useEffect, useRef, useState } from "react";

import { API_BASE } from "./api";

export interface VaultKey {
  key: string;
  source: "env" | "keychain" | "missing";
  available: boolean;
}

export interface VaultStatus {
  ok: boolean;
  count: number;
  keys: VaultKey[];
}

export async function getVaultStatus(): Promise<VaultStatus> {
  const r = await fetch(`${API_BASE}/api/vault/status`);
  if (!r.ok) throw new Error(`HTTP ${r.status}`);
  return r.json() as Promise<VaultStatus>;
}

export function useVaultStatus(intervalMs = 30000): {
  status: VaultStatus | null;
  loading: boolean;
  error: string | null;
  refresh: () => Promise<void>;
} {
  const [status, setStatus] = useState<VaultStatus | null>(null);
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState<string | null>(null);
  const cancelled = useRef(false);

  const refresh = useCallback(async () => {
    try {
      const out = await getVaultStatus();
      if (!cancelled.current) {
        setStatus(out);
        setError(null);
      }
    } catch (e) {
      if (!cancelled.current) setError((e as Error).message);
    } finally {
      if (!cancelled.current) setLoading(false);
    }
  }, []);

  useEffect(() => {
    cancelled.current = false;
    void refresh();
    if (intervalMs > 0) {
      const id = window.setInterval(refresh, intervalMs);
      return () => {
        cancelled.current = true;
        window.clearInterval(id);
      };
    }
    return () => {
      cancelled.current = true;
    };
  }, [refresh, intervalMs]);

  return { status, loading, error, refresh };
}
